File: service/lib/agama/storage/config_conversions/from_model_conversions/drive.rb
# frozen_string_literal: true

# Copyright (c) [2024-2025] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/storage/config_conversions/from_model_conversions/base"
require "agama/storage/config_conversions/from_model_conversions/with_encryption"
require "agama/storage/config_conversions/from_model_conversions/with_filesystem"
require "agama/storage/config_conversions/from_model_conversions/with_partitions"
require "agama/storage/config_conversions/from_model_conversions/with_ptable_type"
require "agama/storage/config_conversions/from_model_conversions/with_search"
require "agama/storage/configs/drive"

module Agama
  module Storage
    module ConfigConversions
      module FromModelConversions
        # Drive conversion from model according to the JSON schema.
        class Drive < Base
          include WithEncryption
          include WithFilesystem
          include WithPtableType
          include WithPartitions
          include WithSearch

          # @param model_json [Hash]
          # @param product_config [Agama::Config]
          # @param encryption_model [Hash, nil]
          def initialize(model_json, product_config, encryption_model = nil)
            super(model_json)
            @product_config = product_config
            @encryption_model = encryption_model
          end

        private

          alias_method :drive_model, :model_json

          # @return [Agama::Config]
          attr_reader :product_config

          # @return [Hash, nil]
          attr_reader :encryption_model

          # @see Base
          # @return [Configs::Drive]
          def default_config
            Configs::Drive.new
          end

          # @see Base#conversions
          # @return [Hash]
          def conversions
            {
              search:      convert_search,
              encryption:  convert_encryption,
              filesystem:  convert_filesystem,
              ptable_type: convert_ptable_type,
              partitions:  convert_partitions(encryption_model)
            }
          end
        end
      end
    end
  end
end

File: web/src/hooks/storage/drive.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { useConfigModel } from "~/hooks/model/storage";
import { putStorageModel } from "~/api";
import { addDrive, deleteDrive, switchToDrive } from "~/storage/drive";
import { useModel } from "~/hooks/storage/model";
import type { ConfigModel, Data } from "~/model/storage/config-model";

function useDrive(name: string): ConfigModel.Drive | null {
  const model = useModel();
  const drive = model?.drives?.find((d) => d.name === name);
  return drive || null;
}

type AddDriveFn = (data: Data.Drive) => void;

function useAddDrive(): AddDriveFn {
  const config = useConfigModel();
  return (data: Data.Drive) => {
    putStorageModel(addDrive(config, data));
  };
}

type DeleteDriveFn = (name: string) => void;

function useDeleteDrive(): DeleteDriveFn {
  const config = useConfigModel();
  return (name: string) => {
    putStorageModel(deleteDrive(config, name));
  };
}

type SwitchToDriveFn = (oldName: string, drive: Data.Drive) => void;

function useSwitchToDrive(): SwitchToDriveFn {
  const config = useConfigModel();
  return (oldName: string, drive: Data.Drive) => {
    putStorageModel(switchToDrive(config, oldName, drive));
  };
}

export { useDrive, useAddDrive, useDeleteDrive, useSwitchToDrive };
export type { AddDriveFn, DeleteDriveFn, SwitchToDriveFn };
